delivery/media/watermarking/lib/media-delivery-watermarking.js

```diff
@@ -0,0 +1,2134 @@
+                    prf.reset(), prf.update(u).finish(u);
+                    for (var k = 0; k < len; k++) t[k] ^= u[k];
+                }
+                for (j = 0; j < len && i * len + j < dkLen; j++) dk[i * len + j] = t[j];
+            }
+            for (i = 0; i < len; i++) t[i] = u[i] = 0;
+            for (i = 0; i < 4; i++) ctr[i] = 0;
+            return prf.clean(), dk;
+        }
+        exports.hkdf = hkdf, exports.pbkdf2 = pbkdf2;
+    }(exports);
+    var sha256 = exports.default;
+    for (var k in exports) sha256[k] = exports[k];
+    module.exports = sha256;
+}();
+
+class Watermarking {
+    constructor(wmOptions) {
+        if (this.wmOptions = wmOptions || {
+            tokenType: TokenType.CWT,
+            validateWMClaims: !0
+        }, this.wmOptions.tokenType = void 0 === this.wmOptions.tokenType ? TokenType.CWT : this.wmOptions.tokenType, 
+        this.wmOptions.validateWMClaims = void 0 === this.wmOptions.validateWMClaims || this.wmOptions.validateWMClaims, 
+        ![ TokenType.CWT, TokenType.JWT ].includes(this.wmOptions.tokenType)) throw new Error("Invalid token type! Only cwt or jwt auth tokens are supported!");
+        this.wmOptions.tokenType === TokenType.CWT ? this.cwtValidator = new CWTValidator(wmOptions) : this.jwtValidator = new JWTValidator(wmOptions);
+    }
+    async validateShortToken(authToken, keys) {
+        if ("string" != typeof authToken) throw new Error("Invalid token type, expected string!");
+        if (isEmptyString(authToken)) throw new Error("Token cannot be empty!");
+        if (0 == keys.length || !keys.every((item => "string" == typeof item && item.trim().length > 0))) throw new Error("Invalid keys type, expected array of hex encoded string!");
+        const tokenParts = authToken.split(".");
+        if (4 !== tokenParts.length) throw new Error("Token malformed: invalid short token format, expected token with four sets of base64url encoded fields joined by a period ‘.’!");
+        const cacheKey = await buildKey([ authToken ].concat(keys)), v = Cache.getShortToken(cacheKey), clockTimestamp = Math.floor(Date.now() / 1e3);
+        if (v) {
+            if (clockTimestamp > v.exp + (this.wmOptions.clockTolerance || 0)) throw new Error("Token expired!");
+            return v;
+        }
+        const exp = parseInt(uint8ArrayToHex(base64url.decode(tokenParts[0], "Uint8Array")), 16);
+        if (clockTimestamp > exp + (this.wmOptions.clockTolerance || 0)) throw new Error("Token expired!");
+        const title = base64url.decode(tokenParts[1], "String"), wmid = parseInt(uint8ArrayToHex(base64url.decode(tokenParts[2], "Uint8Array")), 16), signHex = uint8ArrayToHex(base64url.decode(tokenParts[3], "Uint8Array"));
+        let isTokenVerified = !1;
+        for (const k of keys) {
+            const signature = sha256.exports.hmac(base16.decode(k, "Uint8Array"), (new TextEncoder$1).encode(`${tokenParts[0]}.${tokenParts[1]}.${tokenParts[2]}`));
```

Comment:
> `${tokenParts[0]}.${tokenParts[1]}.${tokenParts[2]}`

I don't know how expensive it is to build this string and then GC it, but it would probably be more efficient to take a substring of the original string. That should cut down on copies and GCs.

Reply:
@evan-hughes : ahh sure, will make this change.